Move hal_nxp glue code to zephyr repo

ZhaoxiangJin opened the pull request:
This PR has been implemented:
  1. Move the hal_nxp glue code from the external module repo to the Zephyr repo. 
      Moving the glue layer to the Zephyr repo can reduce the number of repositories that need to be touched when supporting new devices/drives, making it easier for the development process.
  2. Adopt the MCUXSDK NG (next generation) to hal_nxp.
     The MCUXSDK NG introduced CMake to organize files and Kconfig to resolve dependencies. Adopting MCUXSDK NG to hal_nxp will reduce the effort of introducing new drivers and devices to Zephyr, as it can directly leverage the CMake infrastructure in SDK NG.

### Two integration method compare
**Original integration method:**
The original integration method uses CMake to add SDK code to Zephyr, it is based on modified SDK 2.0 driver CMake. When developers need a driver, they first need to add this CMake file first. Take adc16 driver as an example, the adc16 cmake content is like below. 
```
#Description: ADC16 Driver; user_visible: True
include_guard(GLOBAL)
message("driver_adc16 component is included.")

target_sources(${MCUX_SDK_PROJECT_NAME} PRIVATE
    ${CMAKE_CURRENT_LIST_DIR}/fsl_adc16.c
)

target_include_directories(${MCUX_SDK_PROJECT_NAME} PUBLIC
    ${CMAKE_CURRENT_LIST_DIR}/.
)

include(driver_common)
```
When Zephyr needs adc16 driver, a Kconfig configuration ADC_MCUX_ADC16 is enabled, then add the following code to the glue code to include the adc16 driver.
`include_driver_ifdef(CONFIG_ADC_MCUX_ADC16		adc16		driver_adc16)`

**New integration method:**
SDK NG uses Kconfig and CMake to add driver in project, which is similar with Zephyr. The architecture is:
Device’s Kconfig defines IP Kconfig options, which indicates the device supported features. For example, define MCUX_HW_IP_DriverType_ADC16 if the device has ADC16.
Driver’s Kconfig defines two Kconfig options MCUX_HAS_COMPONENT_driver.xxx and MCUX_COMPONENT_driver.xxx. The driver dependency is handled in this layer. For example, fsl_adc16.c needs fsl_common.c, so MCUX_COMPONENT_driver.adc16 selects MCUX_COMPONENT_driver.common.
```
   config MCUX_HAS_COMPONENT_driver.adc16
       bool
       default y if MCUX_HW_IP_DriverType_ADC16
   
   config MCUX_COMPONENT_driver.adc16
       bool "Use driver adc16"
       select MCUX_COMPONENT_driver.common
       depends on MCUX_HAS_COMPONENT_driver.adc16
```

Driver’s CMakeLists.txt adds source code and include path, based on Kconfig selection.
```
   if(CONFIG_MCUX_COMPONENT_driver.adc16)
       mcux_component_version(2.3.0)

       mcux_add_source(SOURCES fsl_adc16.c fsl_adc16.h)

       mcux_add_include(INCLUDES .)

   endif()
```
More details see [CMakeLists.txt](https://github.com/nxp-mcuxpresso/mcuxsdk-core/blob/release/24.12.00/drivers/adc16/CMakeLists.txt) and [Kconfig](https://github.com/nxp-mcuxpresso/mcuxsdk-core/blob/release/24.12.00/drivers/adc16/Kconfig).

The new integration method will only use the SDK NG CMake files. If the Zephyr side needs the adc16 driver, just add the following code to glue. Developers do not need to add the driver side CMake.
```
if (CONFIG_ADC_MCUX_ADC16)
  set(CONFIG_MCUX_COMPONENT_driver.adc16 ON)
endif()
```

### Device enablement
Based on the new integration method, the device enablement in zephyr becomes easier.
Enabling a device requires the device header, device feature header, peripheral header and so on. All of these have been processed in the device CMakeLists,txt of SDK NG  (see https://github.com/nxp-mcuxpresso/mcuxsdk-core/blob/release/24.12.00/devices/arm/device_header.cmake). We no longer need to add various device related cmakes to hal_nxp like the old integration method, we only need to add the following code in glue code (all devices are the same, so it only needs to be done once): 

```
set(CONFIG_MCUX_COMPONENT_device.system ON)
set(CONFIG_MCUX_COMPONENT_device.CMSIS ON)
```

### The glue code folder structure
```
modules/hal_nxp/
├── Kconfig          # Kconfig entry
├── CMakeLists.txt
├── imx
│   ├── CMakeLists.txt         # From original <hal_nxp_repo>/imx/CMakeLists.txt
│   └── Kconfig.imx             # From original <zephyr_repo>/moduels/Kconfig.imx
├── mcux
│   ├── Kconfig.mcux           # From original <zephyr_repo>/moduels/Kconfig.mcux
│   ├── CMakeLists.txt         # From original <hal_nxp_repo>/mcux/CMakeLists.txt
│   ├── hwmv1.cmake         # From original <hal_nxp_repo>/mcux/hwmv1.cmake
│   ├── mcux-sdk
│   │   └── CMakeLists.txt   # From original <hal_nxp_repo>/mcux/hal_nxp.cmake
│   └── mcux-sdk-ng
│       ├── CMakeLists.txt  #sdk_ng cmake entry
│       ├── components
│       │   └── CMakeLists.txt #sdk_ng component cmake file. Pull SDK component based on requirements and configuration.
│       ├── device
│       │   └── CMakeLists.txt #sdk_ng device cmake file. Pull SDK device based on requirements and configuration.
│       ├── drivers
│       │   └── CMakeLists.txt #sdk_ng driver cmake file. Pull SDK driver based on requirements and configuration.
│       └── middleware
│           ├── bt_controller.cmake   #From original <zephyr_repo>/moduels/hal_nxp/bt_controller/CMakeLists.txt
│           ├── CMakeLists.txt #sdk_ng middleware cmake file. Pull SDK middleware based on requirements and configuration.
│           └── usb_device_config.h  #From original <zephyr_repo>/modules/hal_nxp/usb/usb_device_config.h
└── s32
    ├── CMakeLists.txt       # From original <hal_nxp_repo>/s32/CMakeLists.txt
    └── Kconfig.nxp_s32    # From original <zephyr_repo>/moduels/Kconfig.nxp_s32
```

### The hal_nxp code folder structure
The legacy sdk folder mcux-sdk is not changed, some files in this folder can be removed, such as cmsis_driver, and the devices which are supported by SDK NG.
The mcu-sdk-ng is copied from mcu-sdk, unused files are removed, such as documents, linker script and so on.
The mcux/middleware are improved so that they can be used by both legacy platforms and SDK NG platforms. If it is not necessary to support the legacy platforms, then SDK NG middleware in mcux/mcux-sdk-ng/middleware can be used.
**Note:** Devices supported in SDK NG may also use files in mcux-sdk/components, as these components have files that are not supported in SDK NG, such as rpmsg/fsl_adapter_rfimu.c.
```
mcux/
├── mcux-sdk     # For legacy platforms
│   ├── components
│   ├── devices
│   ├── drivers
│   ├── middleware
│   └── platform
├── mcux-sdk-ng  # For SDK NG platforms
│   ├── arch
│   ├── cmake
│   ├── components
│   ├── devices
│   ├── drivers
│   ├── middleware # SDK NG middleware
│   └── README
├── middleware # Middlewares support both legacy and NG platforms
│   ├── mcux-sdk-middleware-connectivity-framework
│   ├── mcux-sdk-middleware-multicore
│   ├── mcux-sdk-middleware-usb
│   ├── wifi_nxp
│   └── wireless
```

### Why we still need the original hal_nxp.cmake (now is <zephyr_repo>/modules/hal_nxp/mcux/mcux-sdk/CMakeLists.txt)?
The SDK NG does not support legacy devices like MK64F12, MKE16F16 and some APs like MIMX9352_A55 and MIMX8UD7_ADSP. So, these devices still need to be run through hal_nxp.cmake. 
A simple load structure diagram is shown below:
![image](https://github.com/user-attachments/assets/ab83d4e7-fb03-4fd9-af75-30ac8c014b63)

dleach02 commented:
@ZhaoxiangJin please ensure the PR is rebased

butok commented:
> The SDK NG does not support legacy devices like MK64F12, MKE16F16 and some APs like MIMX9352_A55 and MIMX8UD7_ADSP.

Having two HALs at the same time can be confusing.
Can we add legacy devices to the NG SDK and say goodbye to the old HAL?

decsny commented on soc/nxp/imxrt/imxrt7xx/cm33/flash_clock_setup.c:
```diff
@@ -88,12 +88,9 @@ static void flash_init(XSPI_Type *base, CACHE64_CTRL_Type *cache)
-			FSL_FEATURE_XSPI_DLL_REF_VALUE_AUTOUPDATE_X16_DISABLED_REF_COUNTER) |
-		XSPI_DLLCR_SLV_FINE_OFFSET(0) | XSPI_DLLCR_SLV_DLY_OFFSET(0) |
-		XSPI_DLLCR_FREQEN(1U);
+	base->DLLCR[0] |= XSPI_DLLCR_DLLRES(FSL_FEATURE_XSPI_DLL_REF_VALUE_AUTOUPDATE_RES) |
+		XSPI_DLLCR_DLL_REFCNTR(2U) | XSPI_DLLCR_DLL_CDL8(1U) | XSPI_DLLCR_FREQEN(1U) |
+		XSPI_DLLCR_SLV_FINE_OFFSET(0) | XSPI_DLLCR_SLV_DLY_OFFSET(0);
```
is this change/commit bisectable ?

ZhaoxiangJin replied:
no, this change should merge into glue layer update commit

decsny commented on soc/nxp/imxrt/imxrt11xx/power.c:
```diff
@@ -57,7 +57,6 @@ static void gpc_set_transition_flow(void)
 	gpc_tran_step_config_t step_cfg;
 
 	step_cfg.enableStep = true;
-	step_cfg.cntMode = kGPC_StepCounterDisableMode;
```
is this change/commit bisectable ?

ZhaoxiangJin replied:
it is bisectable.

decsny commented on soc/nxp/imxrt/imxrt11xx/soc.c:
```diff
@@ -127,7 +127,7 @@ usb_phy_config_struct_t usbPhyConfig = {
 #endif
 
-#ifdef CONFIG_NXP_IMXRT_BOOT_HEADER
+#if defined(CONFIG_NXP_IMXRT_BOOT_HEADER) && defined(CONFIG_CPU_CORTEX_M7)
```
some of these changes like this one seem like it should be on a separate PR

ZhaoxiangJin replied:
This change should merge into the glue layer update commit

decsny commented on soc/nxp/imxrt/imxrt118x/soc.c:
```diff
@@ -352,7 +352,7 @@ __weak void clock_init(void)
 #endif
 
-#endif /* CONFIG_MCUX_ACMP */
+#endif /* CONFIG_COMPARATOR_MCUX_ACMP || CONFIG_SENSOR_MCUX_ACMP */
```
is this fixing a bug? seems also like it should be a separate PR

ZhaoxiangJin replied:
it is not a bug fix, I removed the kconfig option MCUX_ACMP in the new glue layer, so I update the #if condition here.

decsny commented on drivers/dma/dma_mcux_smartdma.c:
```diff
@@ -13,7 +13,6 @@
 #include <soc.h>
 #include <fsl_smartdma.h>
 #include <fsl_inputmux.h>
-#include <fsl_power.h>
```
in this commit how can you just remove a header file and that's the only change? is this bisectible?

ZhaoxiangJin replied:
It is bisectible. The fsl_power.h is not necessary for the smartdma, and is unavaiable for MCXN in MCUX SDK Next, so, need remove it.

decsny commented on drivers/sensor/nxp/mcux_acmp/Kconfig:
```diff
@@ -9,7 +9,6 @@ config SENSOR_MCUX_ACMP
 	default y
 	depends on DT_HAS_NXP_KINETIS_ACMP_ENABLED
 	select PINCTRL
-	select MCUX_ACMP
```
this should go in the same commit as the other change related to this KConfig names, why do you have two commits

ZhaoxiangJin replied:
yes, this commit should be squashed.

decsny commented on drivers/comparator/Kconfig.mcux_acmp:
```diff
@@ -6,4 +6,3 @@ config COMPARATOR_MCUX_ACMP
 	default y
 	depends on DT_HAS_NXP_KINETIS_ACMP_ENABLED
 	select PINCTRL
-	select MCUX_ACMP
```
this is a third commit where we are just handling this kconfig name, they should all be on one commit

ZhaoxiangJin replied:
this commit should be squashed.

decsny commented on MAINTAINERS.yml:
```diff
@@ -4998,9 +4998,6 @@ West:
     - bperseghetti
   files:
     - modules/hal_nxp/
-    - modules/Kconfig.imx
-    - modules/Kconfig.mcux
-    - modules/Kconfig.nxp_s32
```
okay, I am 100% positive that the commit with these deletions is breaking bisectibility

dleach02 replied:
This particular change probably doesn't break bisectability since this just drives tooling at the github level. But the other commits do need attention.

decsny replied:
Oh woops I think I commented on the wrong thing or misread something, yeah this is not relevant

decsny reviewed (changes requested):
i did a preliminary review of the PR message and the first 9 commits before reviewing the main glue logic change commit, I think there are some problems with the 9 commits breaking bisectibility and some things seem they should be fixed sooner than later in other PRs and not strictly needed only by any of the SDK NG changes. We dont have to split them into separate PRs but we at least cant break bisectibility

dleach02 commented:
> > The SDK NG does not support legacy devices like MK64F12, MKE16F16 and some APs like MIMX9352_A55 and MIMX8UD7_ADSP.
> 
> Having two HALs at the same time can be confusing. Can we add legacy devices to the NG SDK and say goodbye to the old HAL?

I generally agree. But the first step was to transition the HAL to using NG-SDK as it is now. I did ask that we remove the supported NG SOCs from the legacy side for some clarity and then over time, if the SDK team picks up more of the legacy parts into NG we can deprecate them.

dleach02 commented:
@ZhaoxiangJin bisectability is a hard rule for Zephyr Project. Please make a pass compressing the relevant commits to ensure we can build and run across all changes.

zejiang0jason commented:
> > ```
> >    config MCUX_HAS_COMPONENT_driver.adc16
> >        bool
> >        default y if MCUX_HW_IP_DriverType_ADC16
> >    
> >    config MCUX_COMPONENT_driver.adc16
> >        bool "Use driver adc16"
> >        select MCUX_COMPONENT_driver.common
> >        depends on MCUX_HAS_COMPONENT_driver.adc16
> > ```
> 
> Is MCUX_HAS_COMPONENT_driver.adc16 always equal to MCUX_HW_IP_DriverType_ADC16? Can we reduce it to:
> 
> ```
>     config MCUX_COMPONENT_driver.adc16
>         bool "Use driver adc16"
>         select MCUX_COMPONENT_driver.common
>         depends on MCUX_HW_IP_DriverType_ADC16
> ```

Most times yes, sometimes it is not one-to-one map, e.g., one driver works for different IPs, one driver works with a combination of several IPs...

ZhaoxiangJin committed:
modules: hal_nxp: Move hal_nxp glue layer to zephyr repo

Move hal_nxp glue layer to zephyr repo.
Fix build warnings and failures caused by hal_nxp upgrade.
Update manifest to contain hal_nxp changes.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
> @ZhaoxiangJin bisectability is a hard rule for Zephyr Project. Please make a pass compressing the relevant commits to ensure we can build and run across all changes.

Yes, some of the changes 

> @ZhaoxiangJin bisectability is a hard rule for Zephyr Project. Please make a pass compressing the relevant commits to ensure we can build and run across all changes.

ZhaoxiangJin commented:
> @ZhaoxiangJin bisectability is a hard rule for Zephyr Project. Please make a pass compressing the relevant commits to ensure we can build and run across all changes.

Yes, I have paid attention to maintaining bisectability during development, but [ this commit](https://github.com/zephyrproject-rtos/zephyr/pull/84423/commits/e75538ead46f478e692fb986369cb5ad8d4b7f5d) still breaks bisectability. This commit should be put into the [glue code update commit](https://github.com/zephyrproject-rtos/zephyr/pull/84423/commits/211ed3aa2422f7b1e6b767ef4562d1fa5c3ec151) to avoid regression problems. In addition, commits [1](https://github.com/zephyrproject-rtos/zephyr/pull/84423/commits/f5b1887d627e5252b6a1c6b6e9f6c5e100c19823), [2](https://github.com/zephyrproject-rtos/zephyr/pull/84423/commits/a09f224efa7d50e595fc1ae3a98f6bf7d47e864a), and [3](https://github.com/zephyrproject-rtos/zephyr/pull/84423/commits/1da23445b2d09b911a859965935eea4bcec45295) should be compressed into one commit.

hakehuang commented:
build is OK, but mimxrt118_evk still has board testing issue, in checking.

ZhaoxiangJin commented:
> build is OK, but mimxrt118_evk still has board testing issue, in checking.

The test case tests/subsys/canbus/isotp/conformance works good on my local test. The run log attached:
```
*** Booting Zephyr OS build v4.1.0-686-g147893a84e35 ***
Running TESTSUITE isotp_conformance
===================================================================
START - test_canfd_mandatory_padding
SKIP - test_canfd_mandatory_padding in 0.001 seconds
===================================================================
START - test_canfd_rx_dl_validation
SKIP - test_canfd_rx_dl_validation in 0.001 seconds
===================================================================
START - test_receive_data
PASS - test_receive_data in 0.295 seconds
===================================================================
START - test_receive_data_blocks
PASS - test_receive_data_blocks in 1.149 seconds
===================================================================
START - test_receive_sf
I: SF DL does not fit. Ignore
PASS - test_receive_sf in 0.207 seconds
===================================================================
START - test_receive_sf_ext
I: SF DL does not fit. Ignore
PASS - test_receive_sf_ext in 0.207 seconds
===================================================================
START - test_receive_sf_fixed
PASS - test_receive_sf_fixed in 0.207 seconds
===================================================================
START - test_receive_timeouts
E: Timeout while waiting for CF
PASS - test_receive_timeouts in 1.006 seconds
===================================================================
START - test_receiver_fc_errors
E: Sequence number mismatch
I: Got unexpected frame. Ignore
I: Got unexpected frame. Ignore
I: Got unexpected frame. Ignore
I: Got unexpected frame. Ignore
I: Got unexpected frame. Ignore
I: Got unexpected frame. Ignore
E: Pkt length is 4089 but buffer has only 768 bytes
PASS - test_receiver_fc_errors in 0.640 seconds
===================================================================
START - test_send_data
PASS - test_send_data in 0.448 seconds
===================================================================
START - test_send_data_blocks
PASS - test_send_data_blocks in 1.000 seconds
===================================================================
START - test_send_sf
PASS - test_send_sf in 0.202 seconds
===================================================================
START - test_send_sf_ext
PASS - test_send_sf_ext in 0.202 seconds
===================================================================
START - test_send_sf_fixed
PASS - test_send_sf_fixed in 0.203 seconds
===================================================================
START - test_send_timeouts
E: Reception of next FC has timed out
E: Reception of next FC has timed out
E: Reception of next FC has timed out
PASS - test_send_timeouts in 3.828 seconds
===================================================================
START - test_sender_fc_errors
E: Got overflow FC frame
I: Got to many wait frames
PASS - test_sender_fc_errors in 0.624 seconds
===================================================================
START - test_stmin
PASS - test_stmin in 0.266 seconds
===================================================================
TESTSUITE isotp_conformance succeeded
Running TESTSUITE isotp_conformance_mode_check
===================================================================
START - test_bind
PASS - test_bind in 0.001 seconds
===================================================================
START - test_send
PASS - test_send in 0.002 seconds
===================================================================
TESTSUITE isotp_conformance_mode_check succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [isotp_conformance]: pass = 15, fail = 0, skip = 2, total = 17 duration = 10.486 seconds
- SKIP - [isotp_conformance.test_canfd_mandatory_padding] duration = 0.001 seconds
- SKIP - [isotp_conformance.test_canfd_rx_dl_validation] duration = 0.001 seconds
- PASS - [isotp_conformance.test_receive_data] duration = 0.295 seconds
- PASS - [isotp_conformance.test_receive_data_blocks] duration = 1.149 seconds
- PASS - [isotp_conformance.test_receive_sf] duration = 0.207 seconds
- PASS - [isotp_conformance.test_receive_sf_ext] duration = 0.207 seconds
- PASS - [isotp_conformance.test_receive_sf_fixed] duration = 0.207 seconds
- PASS - [isotp_conformance.test_receive_timeouts] duration = 1.006 seconds
- PASS - [isotp_conformance.test_receiver_fc_errors] duration = 0.640 seconds
- PASS - [isotp_conformance.test_send_data] duration = 0.448 seconds
- PASS - [isotp_conformance.test_send_data_blocks] duration = 1.000 seconds
- PASS - [isotp_conformance.test_send_sf] duration = 0.202 seconds
- PASS - [isotp_conformance.test_send_sf_ext] duration = 0.202 seconds
- PASS - [isotp_conformance.test_send_sf_fixed] duration = 0.203 seconds
- PASS - [isotp_conformance.test_send_timeouts] duration = 3.828 seconds
- PASS - [isotp_conformance.test_sender_fc_errors] duration = 0.624 seconds
- PASS - [isotp_conformance.test_stmin] duration = 0.266 seconds

SUITE PASS - 100.00% [isotp_conformance_mode_check]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.003 seconds
- PASS - [isotp_conformance_mode_check.test_bind] duration = 0.001 seconds
- PASS - [isotp_conformance_mode_check.test_send] duration = 0.002 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

hakehuang commented:
@ZhaoxiangJin I pull the PR again and re-run mimxrt1180_evk, now seems ok.

dleach02 commented:
@hakehuang can you provide confirmation on our test status on the board farm for this PR?

dleach02 commented:
@decsny the commits have been squashed to address the bisectability

hakehuang commented:
board testing and all NXP platform build pass on v4.1.0-983-g73bc86c241d7
